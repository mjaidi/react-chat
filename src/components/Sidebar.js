import React, { Component } from "react";
import { Subscription } from "react-apollo";
import gql from "graphql-tag";
import Loader from "components/Loader";

export const fetchMyConversations = gql`
  subscription fetchMyConversations($id: String!) {
    conversations(where: { conversation_members: { user_id: { _eq: $id } } }) {
      id
      conversation_members {
        id
        member {
          id
          auth0_id
          username
        }
      }
    }
  }
`;
export const fetchOtherUsers = gql`
  subscription fetchOtherUsers($id: String!) {
    users(
      where: {
        _not: {
          user_conversations: {
            conversation: {
              conversation_members: { member: { auth0_id: { _eq: $id } } }
            }
          }
        }
      }
    ) {
      id
      auth0_id
      username
    }
  }
`;

export const newConversation = gql`
  mutation newConversation($currentUser: String!, $otherUser: String!) {
    insert_conversation_members(
      objects: {
        conversation: {
          data: { conversation_members: { data: { user_id: $otherUser } } }
        }
        user_id: $currentUser
      }
    ) {
      returning {
        conversation {
          id
        }
      }
    }
  }
`;

class Sidebar extends Component {
  goTo(route, id) {
    this.props.history.push({
      pathname: `/${route}/${id}`,
      id: id
    });
  }

  handleNewConversation(e, otherUser, currentUser, mutate) {
    e.preventDefault();
    mutate({
      mutation: newConversation,
      variables: {
        currentUser: currentUser,
        otherUser: otherUser
      }
    }).then(result => {
      this.goTo(
        "home/conversations",
        result.data.insert_conversation_members.returning[0].conversation.id
      );
    });
  }

  conversationMemberNames(conversation, currentUser) {
    // Get only the usernames of the other participants of a conversation
    const otherMembers = conversation.conversation_members.filter(
      m => m.member.auth0_id !== currentUser
    );
    return otherMembers.map(m => m.member.username).join(", ");
  }

  render() {
    const currentUser = localStorage.getItem("auth0:id_token:sub");
    return (
      <aside id="sidebar">
        <Subscription
          subscription={fetchMyConversations}
          variables={{ id: currentUser }}
        >
          {({ data, error, loading }) => {
            if (loading) {
              return <Loader />;
            }
            if (error) {
              return "Error loading conversations";
            }
            return (
              <div>
                <p className="sidebar-heading">
                  {" "}
                  My Conversations (
                  {!data.conversations ? 0 : data.conversations.length})
                </p>
                <ul className="sidebar-list">
                  {data.conversations.map(c => {
                    return (
                      <li
                        key={c.id}
                        className={
                          c.id === this.props.activeConversation
                            ? "active"
                            : null
                        }
                        onClick={this.goTo.bind(
                          this,
                          "home/conversations",
                          c.id
                        )}
                      >
                        {this.conversationMemberNames(c, currentUser)}
                      </li>
                    );
                  })}
                </ul>
              </div>
            );
          }}
        </Subscription>
        <Subscription
          subscription={fetchOtherUsers}
          variables={{ id: currentUser }}
        >
          {({ data, error, loading }) => {
            if (loading) {
              return <Loader />;
            }
            if (error) {
              return "Error loading  users";
            }
            const filteredUsers = data.users.filter(
              u => u.auth0_id !== currentUser
            );
            return (
              <div>
                <p className="sidebar-heading"> Chat with other users</p>
                <ul className="sidebar-list">
                  {filteredUsers.map(u => {
                    return (
                      <li
                        key={u.id}
                        onClick={e =>
                          this.handleNewConversation(
                            e,
                            u.auth0_id,
                            currentUser,
                            this.props.client.mutate
                          )
                        }
                      >
                        {u.username}
                      </li>
                    );
                  })}
                </ul>
              </div>
            );
          }}
        </Subscription>
      </aside>
    );
  }
}

export default Sidebar;
