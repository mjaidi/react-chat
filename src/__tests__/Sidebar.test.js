import React from "react";
import { mount } from "enzyme";
import { MockedProvider, wait } from "@apollo/react-testing";

import Sidebar, { fetchMyConversations } from "components/Sidebar";

const mocks = [
  {
    request: {
      query: fetchMyConversations,
      variables: {
        id: "1341"
      }
    },
    result: {
      data: {
        conversations: [
          {
            id: "1",
            conversation_members: [
              {
                id: "1",
                member: {
                  id: "1",
                  auth0_id: "1",
                  username: "majid",
                  __typename: "users"
                },
                __typename: "conversation_members"
              }
            ],
            __typename: "conversations"
          }
        ]
      }
    }
  }
];

describe("Sidebar", () => {
  let component;
  const wait = (time = 0) => new Promise(res => setTimeout(res, time));
  const executeMockProviderTestCase = componentInstance => {
    return wait(100).then(() => componentInstance.update());
  };
  beforeEach(() => {
    component = mount(
      <MockedProvider mocks={mocks} addTypeName={false} removeTypename>
        <Sidebar />
      </MockedProvider>
    );
  });
  it("should render correctly without crashing", () => {
    expect(component).toMatchSnapshot();
  });

  it("should contain 1 conversation", () => {
    return executeMockProviderTestCase(component).then(() => {
      expect(component.find(".sidebar-link").length).to.equal(1);
    });
  });
});
