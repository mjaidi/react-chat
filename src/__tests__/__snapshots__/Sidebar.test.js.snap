// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should render correctly without crashing 1`] = `undefined`;
